cs/codec-template.cs.j2

```diff
@@ -0,0 +1,266 @@
+            response.{{ param.name|capital }} = {{ decode_var_sized(param) }};
+        {% endfor %}
+            return response;
+        }
+    {# EVENTS#}
+    {% if method.events|length != 0 %}
+    {% for event in method.events%}
+    
+        public static ClientMessage Encode{{ event.name|capital }}Event({% for param in event.params %}{{ lang_types_encode(param.type) }} {{param.name}}{% if not loop.last %}, {% endif %}{% endfor %}) 
+        {
+            var clientMessage = CreateForEncode();
+            var initialFrame = new Frame(new byte[Event{{ event.name}}InitialFrameSize], UnfragmentedMessage);
+            initialFrame.Flags |= IsEventFlag;
+            EncodeInt(initialFrame.Content, TypeFieldOffset, Event{{ (event.name)}}MessageType);
+        {% for param in fixed_params(event.params) %}
+            Encode{{ lang_types_decode(param.type)|capital }}(initialFrame.Content, Event{{ (event.name)}}{{(param.name)}}FieldOffset, {{ param.name }});
+        {% endfor %}
+            clientMessage.Add(initialFrame);
+        {% for param in var_size_params(event.params) %}
+            {{ encode_var_sized(param) }};
+        {% endfor %}
+            return clientMessage;
+        }
+        {% endfor %}
+
+        public abstract class AbstractEventHandler 
+        {
+            public void Handle(ClientMessage clientMessage) 
+            {
+                var messageType = clientMessage.MessageType;
+                var iterator = clientMessage.GetIterator();
+            {% for event in method.events%}
+                if (messageType == Event{{(event.name)}}MessageType) {
+                {% if  fixed_params(event.params)|length != 0 %}
+                    var initialFrame = iterator.Next();
+                {% else %}
+                    //empty initial frame
+                    iterator.Next();
+                {% endif %}
+                {% for param in fixed_params(event.params) %}
+                    {{ lang_types_decode(param.type) }} {{param.name}} =  Decode{{ lang_types_decode(param.type)|capital }}(initialFrame.Content, Event{{(event.name)}}{{param.name}}FieldOffset);
+                {% endfor %}
+                {% for param in var_size_params(event.params) %}
+                    {{ lang_types_decode(param.type) }} {{param.name}} = {{ decode_var_sized(param) }};
+                {% endfor %}
+                    Handle{{ event.name|capital }}Event({% for param in event.params %}{{param.name}}{% if not loop.last %}, {% endif %}{% endfor %});
+                    return;
+                }
+            {% endfor %}
+                Logger.GetLogger(GetType()).Finest("Unknown message type received on event handler :" + messageType);
+            }
+        {% for event in method.events%}
+
+            public abstract void Handle{{ event.name|capital }}Event({% for param in event.params %}{{ lang_types_encode(param.type) }} {{param.name}}{% if not loop.last %}, {% endif %}{% endfor %});
```

Comment:
old protocol generator was using delegates, instead of abstract methods. overriding these methods on client code was not clean wrt passing method as parameters.